include/dma.h

```diff
@@ -35,114 +35,96 @@ enum DMAEvent {
 	DMA_UNMASKED,
 };
 
+class Section;
+using DMA_ReservationCallback = std::function<void(Section*)>;
+
 class DmaChannel;
-using DMA_CallBack = std::function<void(DmaChannel *chan, DMAEvent event)>;
+using DMA_Callback = std::function<void(DmaChannel *chan, DMAEvent event)>;
 
 class DmaChannel {
 public:
-	uint32_t pagebase;
-	uint16_t baseaddr;
-	uint32_t curraddr;
-	uint16_t basecnt;
-	uint16_t currcnt;
-	uint8_t channum;
-	uint8_t pagenum;
-	uint8_t DMA16;
-	bool increment;
-	bool autoinit;
-//	uint8_t trantype; //Not used at the moment
-	bool masked;
-	bool tcount;
-	bool request;
-	DMA_CallBack callback;
+	uint32_t pagebase = 0;
+	uint32_t curraddr = 0;
 
-	DmaChannel(uint8_t num, bool dma16);
+	uint16_t baseaddr = 0;
+	uint16_t basecnt  = 0;
+	uint16_t currcnt  = 0;
+
+	uint8_t channum = 0;
+	uint8_t pagenum = 0;
+	uint8_t DMA16   = 0;
+
+	bool increment = false;
+	bool autoinit  = false;
+	bool masked    = true;
+	bool tcount    = false;
+	bool request   = false;
 
-	void DoCallBack(DMAEvent event) {
-		if (callback)
-			callback(this, event);
-	}
-	void SetMask(bool _mask) {
-		masked=_mask;
-		DoCallBack(masked ? DMA_MASKED : DMA_UNMASKED);
-	}
-	void Register_Callback(DMA_CallBack _cb) { 
-		callback = _cb; 
-		SetMask(masked);
-		if (callback) Raise_Request();
-		else Clear_Request();
-	}
-	void ReachedTC(void) {
-		tcount=true;
-		DoCallBack(DMA_REACHED_TC);
-	}
-	void SetPage(uint8_t val) {
-		pagenum=val;
-		pagebase=(pagenum >> DMA16) << (16+DMA16);
-	}
-	void Raise_Request(void) {
-		request=true;
-	}
-	void Clear_Request(void) {
-		request=false;
-	}
-	size_t Read(size_t words, uint8_t *dest_buffer)
-	{
-		return ReadOrWrite(DMA_DIRECTION::READ, words, dest_buffer);
-	}
-	size_t Write(size_t words, uint8_t *src_buffer)
-	{
-		return ReadOrWrite(DMA_DIRECTION::WRITE, words, src_buffer);
-	}
+	DMA_Callback callback = {};
+
+	DmaChannel(uint8_t num, bool dma16);
+	~DmaChannel();
+
+	void DoCallback(DMAEvent event);
+	void SetMask(bool _mask);
+	void Register_Callback(const DMA_Callback cb);
+	void ReachedTC();
+	void SetPage(uint8_t val);
+	void Raise_Request();
+	void Clear_Request();
+	size_t Read(size_t words, uint8_t* dest_buffer);
```

Comment:
Mark method as const?

Reply:
Marked the buffer as a const-pointer, but the Channel itself is modified during the read (it keeps track of how much data is left before reaching terminal count).